### PR TITLE
Negotiate amount of cargo

### PR DESCRIPTION
**Pull request by robothauler**
Requested by @impaktor some time ago for the cargo run module.
Player with smaller ships can now choose a smaller amount of cargo. The advertisement remain in the BB with the rest of the cargo.

### CODE REVIEW

File: data/lang/module-cargorun/en.json
Comment (impaktor) [inline data/lang/module-cargorun/en.json:232]: Just a personal opinion, I think HOWMUCH_4 and HOWMUCH_2 are basically the same, i.e. a player seeing two adverts, with each flavor will probably not notice they're technically different. (Generally, I'd opt for having fewer flavours if they're this similar)
Action: none required, I just felt like pointing out my view (which could be wrong) on this subject.

File: data/lang/module-cargorun/en.json
Comment (impaktor) [inline data/lang/module-cargorun/en.json:248]: I haven't playtested, and maybe this is needed for the code to enter a menu system, but if not, one could simply have the player, instead of asking to negotiate, make a counter-deal right away: 
"Unfortunately I can only take:
- [x] 10t
- [ ] 100t 
- [ ] 1000t"

File: data/modules/CargoRun/CargoRun.lua
Comment (impaktor) [inline data/modules/CargoRun/CargoRun.lua:385]: In general: using the same string twice "HOWMUCH_" and "HOWMUCH", I'm half thinking if this might be made more generic, but I'm not suggesting any change. Just thinking out loud.

**Review by impaktor — APPROVED** [review]
Cool. I forgot about this PR (sorry).
I haven't tested it, but I trust it's good to go. 
My comments are just me thinking out loud, if they inspire you to want to make any changes, then notify us so we don't merge it before you've had a chance to update the PR. If not, we'll merge it as it is, next time someone with merge-power feels like merging PRs.

Thanks!

File: data/lang/module-cargorun/en.json
Comment (robothauler) [reply data/lang/module-cargorun/en.json:232]: Yes, sounds very similar. I will remove one.

File: data/lang/module-cargorun/en.json
Comment (robothauler) [reply data/lang/module-cargorun/en.json:248]: Yes, it is needed to enter the menu with the amounts.

**Comment by impaktor** [comment]
I'm just trying this out now.

I wonder if it should be more clearly worded that payment scales linearly (baring round off) with mass? It's easy to see in the below example since it's multiples of $25, which is easy to do head math on. 

2021-05-23-140302_1600x900_scrot

But perhaps one would like the text to read: "Yes we can split the cargo. I have 9t in total, I'll pay $25 for each tonne you can transport for me"

This is an open question, as it's not obvious how much we should spell out to the player. Perhaps we _want_ them to do some thinking and actually consider if it's a good deal or bad deal to take a fraction of the total mass?

Also:
When clicking "transport Xt for $Y", you get the below:

2021-05-23-141255_1600x900_scrot

I assume the "deal" is sealed when I again click "OK, agreed", I wonder if she should summarize the deal a bit more: 
"OK, agreed: you will take Xt for $Y?"

I haven't looked at the code in detail, but they only refuse to split the cargo when it's 1t? If so, there could be those who prefer a "fill or kill" deal, i.e. you either take it all or none of it. I'm just throwing it out there.

Anyway, those are my thoughts when I tried it out, what do you/others think?
We could merge this as is, it depends on if @robothauler agrees/want to work on any of the above points.

**Comment by robothauler** [comment]
To be honest, after a few days of playing around, I'm no longer satisfied with this PR. So I also think this should be improved.

**Comment by impaktor** [comment]
What are you dissatisfied with? Overall I think it's good. One thought I had was if it's possible to not re-create a new advert for the remaining goods left to be shipped, but if it's possible to just update the existing advert, i.e. pick out the relevant parts of the table with the advert info and update, so just lower the remaining cargo to be shipped?

Anyway, I also don't want to stall this from going in, I've been much too slow to review this as it is.

**Comment by robothauler** [comment]
> What are you dissatisfied with? Overall I think it's good. One thought I had was if it's possible to not re-create a new advert for the remaining goods left to be shipped, but if it's possible to just update the existing advert, i.e. pick out the relevant parts of the table with the advert info and update, so just lower the remaining cargo to be shipped?

The recreation with the remaining cargo works. But there are still some oddities.
- The player can gain more reputation by clicking 10 times 1t for a 10t advert
- e.g. the customer says: 1, 2, 4, 6, 9t for selection as in your example above. I have 8t available on my ship. So I click 6t and then 2t. Why not 8t at once?
- The cash difference between a 2t delivery advert and a 10t delivery advert with a 2t selection is too big IMHO

> Anyway, I also don't want to stall this from going in, I've been much too slow to review this as it is.

I also have no problem with a later improvement after merge.

**Comment by impaktor** [comment]
> The player can gain more reputation by clicking 10 times 1t for a 10t advert

Good point. You could add an additional attribute to the advert "denyPlayer = false" or similar, that flips to true if the player takes the mission with "negotiated cargo", and flip it to "true" for the new advert with the remaining cargo, then deny the player snarkily: "you again? We already have a contract, and I don't want to sign more with you flipp flopping", (but differently worded). Or just check that variable when assigning reputation, so only the first version of the advert gives reputation, and the other give 0 reputation.

**Comment by robothauler** [comment]
Feels much better now. I did not touch the reward. I have changed my mind. That should be the advantage of a freighter compared to a courier or fighter where cargo mass is not so important.

**Comment by impaktor** [comment]
Tested it, seems to work as advertised.

For future: could be good to have some boolean `will_split_cargo` variable set to an advert, as some might only want to deal with a single transporter, and as @jimishol pointed out, it gives an incentive to have a larger ship.
But that's for another day/PR.

Thanks for the work!

**Review by impaktor — COMMENTED** [review]
Tested it, seems to work as advertised.

For future: could be good to have some boolean will_split_cargo variable set to an advert, as some might only want to deal with a single transporter, and as @jimishol pointed out, it gives an incentive to have a larger ship.
But that's for another day/PR.

Thanks for the work!

**Comment by impaktor** [comment]
@robothauler Feedback from the [player forum](https://spacesimcentral.com/community/pioneer/graphics-issues/#post-60225):

> I do indeed have a first feedback for you regarding the new feature (negotiating the quantity to be transported). Because the option 'how much to transport' is no longer available, you now have to use the option 'negotiate' to find out how much it is at all. Doesn't make much sense in my opinion. It would be better if it was the same as before and you put the option of negotiating in there?

**Pull request merged**
